### PR TITLE
feat(vpc): additional validation around Subnet Types

### LINKED ISSUE #3704: VPC with only private subnet fails with error
### The Question

Because creating a VPC with default settings creates a public and private subnets that default to a NAT gateway for each public subnet, i am running out of EIP. 
My application doesnt even need a NAT gateways, its just an app that uses dax to connect to dynamodb.
Anyways, i tried creating a vpc with just private subnet and add dynamo as a gateway endpoint:

```
const daxVpc = new Vpc(this, 'daxVpc', {
      cidr: Vpc.DEFAULT_CIDR_RANGE,
      subnetConfiguration: [{
        subnetType: SubnetType.PRIVATE,
        name: defaultSubnetName(SubnetType.PRIVATE)
      }],
      gatewayEndpoints: {
        DYNAMODB: {
          service: GatewayVpcEndpointAwsService.DYNAMODB
        }
      }
    });
```

but it errors out during deployment:

```
8/31 | 16:24:34 | CREATE_IN_PROGRESS   | AWS::EC2::Subnet                      | daxVpc/IsolatedSubnet1/Subnet (daxVpcIsolatedSubnet1SubnetA090FC82) Resource creation Initiated
  8/31 | 16:24:34 | CREATE_IN_PROGRESS   | AWS::EC2::Subnet                      | daxVpc/IsolatedSubnet2/Subnet (daxVpcIsolatedSubnet2Subnet469D29C8) Resource creation Initiated
  9/31 | 16:24:34 | CREATE_COMPLETE      | AWS::EC2::RouteTable                  | daxVpc/IsolatedSubnet2/RouteTable (daxVpcIsolatedSubnet2RouteTableE102F846) 
 10/31 | 16:24:34 | CREATE_COMPLETE      | AWS::EC2::RouteTable                  | daxVpc/IsolatedSubnet1/RouteTable (daxVpcIsolatedSubnet1RouteTable5B90AC7F) 
 11/31 | 16:24:36 | CREATE_FAILED        | AWS::EC2::Route                       | daxVpc/IsolatedSubnet1/DefaultRoute (daxVpcIsolatedSubnet1DefaultRoute54F632D7) Exactly one of [EgressOnlyInternetGatewayId, InstanceId, NatGatewayId, TransitGatewayId, GatewayId, VpcPeeringConnectionId, NetworkInterfaceId] must be specified and not empty
```

What configuration am i missing ??

### Environment

  - **CDK CLI Version:** 1.3.0
  - **Module Version:** 1.3.0
  - **OS:** 
  - **Language:** TypeScript

### PR DESCRIPTION
**Pull request by rix0rrr**
Try to improve the usability around VPCs and certain Subnet Type
configurations.

- Make clear that ISOLATED does not mean "no Internet access at all",
  just no Internet access from the point of view of the current VPC.

- If people configure natGateways=0, tell them that they probably meant
  to configure no PRIVATE subnets.

- If people DO configure PRIVATE subnets, make sure they also configure
  PUBLIC subnets, otherwise we won't be able to place the NAT gateways
  anywhere useful.

- If people end up with a VPC without PRIVATE subnets, the default
  behavior of `selectSubnets` is pretty useless, because it can never
  work, and that's not what people are used to from the CDK. Dynamically
  adjust the selection default to whatever subnet types *are* available.

Fixes #3704.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 1]
```diff
@@ -166,7 +165,7 @@ export interface SubnetSelection {
    *
    * At most one of `subnetType` and `subnetGroupName` can be supplied.
    *
-   * @default SubnetType.PRIVATE
+   * @default SubnetType.PRIVATE (or ISOLATED or PUBLIC if there are no PRIVATE subnets)
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts]: I'm not sure this description of `@default` is super clear... `ISOLATED or PUBLIC`? I don't think this is a valid value... Do you mean whichever value resolves to a subnet first in that order?

Comment (eladb) [reply packages/@aws-cdk/aws-ec2/lib/vpc.ts]: Yes, @rix0rrr I like you proposed change (and the value should be `-`). It should be something like:

```ts
/**
 * @default - the default subnet type will be the first available type in the following order: PRIVATE, ISOLATE, PUBLIC
 */

Something like that.

File: packages/@aws-cdk/aws-ec2/test/test.vpc.ts [hunk 2]
```diff
@@ -415,6 +415,18 @@ export = {
...
       test.done();
     },
+
+    'natGateways = 0 requires there to be no PRIVATE subnets'(test: Test) {
+      const stack = getTestStack();
+      test.throws(() => {
+        new Vpc(stack, 'VPC', {
+          natGateways: 0,
+        });
+      }, /make sure you don't configure any PRIVATE subnets/);
```
Comment (eladb) [inline packages/@aws-cdk/aws-ec2/test/test.vpc.ts]: Perhaps provide the rationale for this in the error message or an accompanying link to an issue or doc

**Comment by ChrisLahaye** [comment]
What about NAT instances (https://docs.aws.amazon.com/vpc/latest/userguide/VPC_NAT_Instance.html)? They have zero NAT gateways and one or more private subnets, therefore failing this validation.

**Pull request merged**